mp/src/game/shared/momentum/run/mom_replay_factory.cpp

```diff
@@ -98,4 +108,201 @@ CMomReplayBase* CMomReplayFactory::LoadReplayFile(const char* pFileName, bool bF
     return toReturn;
 }
 
+bool CMomReplayFactory::IsReplayDebugEnabled()
+{
+#ifdef CLIENT_DLL
+    return mom_replay_debug.GetBool();
+#else
+    static ConVarRef replayDebug("mom_replay_debug");
+    return replayDebug.GetBool();
+#endif
+}
+
+#ifdef CLIENT_DLL
+AsyncReplayStatus_t CMomReplayFactory::AsyncLoadReplaysByFilename(const CUtlStringList &filenames, const char *pathID,
+                                                                 const MomLocalReplayCallback_t &callback,
+                                                                 MomReplayAsyncHandle &outHandle)
+{
+    bool bLogReplay = IsReplayDebugEnabled();
+
+    if (bLogReplay)
+        Log("Loading replays asynchronously...\n");
+
+    int idx = requests.AddToTail();
+    auto &internalRequest = requests[idx];
+
+    // This belongs in a constructor but CUtlVector doesn't have emplace_back
+    size_t count = filenames.Size();
+    internalRequest.vecControls.EnsureCount(count);
+    internalRequest.vecRequests.EnsureCount(count);
+    internalRequest.controls_pending = count;
+
+    outHandle = internalRequest.handle;
+
+    internalRequest.callback = callback;
+
+    FOR_EACH_VEC(filenames, i)
+    {
+        internalRequest.vecRequests[i].pszFilename = filenames[i];
+        internalRequest.vecRequests[i].pfnCallback = FilesystemCallbackForward;
+        internalRequest.vecRequests[i].pszPathID = pathID;
+        internalRequest.vecRequests[i].pContext = internalRequest.handle;
+    }
+
+    if (filesystem->AsyncReadMultiple(internalRequest.vecRequests.Base(), count, internalRequest.vecControls.Base())
+        != FSASYNC_OK)
+    {
+        Warning("Bad async replay load request!");
+        AsyncCleanup(internalRequest.handle);
+        return ASYNC_REPLAYS_FSERROR;
+    }
+
+    return ASYNC_REPLAYS_INPROGRESS;
+
+}
+
+AsyncReplayStatus_t CMomReplayFactory::AsyncLoadReplaysForMap(const char *mapName,
+                                                              const MomLocalReplayCallback_t &callback,
+                                                              MomReplayAsyncHandle &outHandle)
+{
+    char path[MAX_PATH];
+    Q_snprintf(path, MAX_PATH, "%s/%s-*%s", RECORDING_PATH, mapName, EXT_RECORDING_FILE);
+    Q_FixSlashes(path);
+
+    FileFindHandle_t found;
+    CUtlStringList filenames;
+    const char *pFoundFile = filesystem->FindFirstEx(path, "MOD", &found);
+
+    if (!pFoundFile)
+    {
+        filesystem->FindClose(found);
+        return ASYNC_REPLAYS_NONEFOUND;
+    }
+
+    // NOTE: THIS NEEDS TO BE MANUALLY CLEANED UP!
+    char pReplayPath[MAX_PATH];
+
+    do
+    {
+        // This is not 100% foolproof validation but it should be enough unless you purposefully break it
+        if (Q_strlen(pFoundFile) == Q_strlen(mapName) + 1 /* - */ + REPLAY_HASH_LENGTH + Q_strlen(EXT_RECORDING_FILE))
```

Comment:
Okay so after a bit of reading the original code I understand why it breaks. The `ConvertLocalTimes` guy automatically assumes that the filename of the read replay *had* to be `mapname-hash.mrf`, but isn't always the case with the files you were messing around with.

Unfortunately as I found out, downloaded online replays have the format `mapname-runID.mrf` with file name lengths substantially smaller than the SHA1 hash ones.

So here's what I propose. Inside of `CMomReplayBase` we add a new variable for the file path it was read from. So far our use case for replays is strictly file-based, even ones from online leaderboards are locally downloaded as to save our servers the bandwidth. This can be a `char m_szFilePath[MAX_PATH];` right below the hash of the replay in the private section. Then, you would want to expose a public virtual getter and setter for this file path.

Then, for everything that loads a replay from file, you can do something like `replay.SetFilePath(filepath);`. For this case, it'd be in `FileSystemCallback`, right before it gets added to the vector, and you can get the file name from the `FileAsyncRequest_t` object.

This can then let us get the exact file name of the replay from the pointer of it, and thus it doesn't matter what it's called. The `ConvertLocalTimes` guy would instead use the public getter from the pointer instead of reconstructing it and making assumptions. You can also use it directly in my previous suggestion with regards to the `ContextDeleteReplay` stuff. You can also then safely remove this check as it won't matter what it's called.